File: pdu/ruby/pdu_runtime.rb
module HakoPdu
  module Runtime
    PDU_META_DATA_SIZE = 24
    PDU_META_DATA_MAGICNO = 0x12345678
    PDU_META_DATA_VERSION = 1

    module PduMetaDataState
      INVALID = 0
      UNINITIALIZED = 1
      VALID = 2
    end

    def self.classify_pdu_metadata(binary_data)
      return PduMetaDataState::INVALID if binary_data.nil?

      bytes = binary_data.b
      return PduMetaDataState::INVALID if bytes.bytesize < PDU_META_DATA_SIZE
      header = bytes.byteslice(0, PDU_META_DATA_SIZE)
      return PduMetaDataState::UNINITIALIZED if header == "\x00".b * PDU_META_DATA_SIZE

      magic, version = header.unpack('V2')
      if magic == PDU_META_DATA_MAGICNO && version == PDU_META_DATA_VERSION
        PduMetaDataState::VALID
      else
        PduMetaDataState::INVALID
      end
    end
  end
end

File: pdu/elixir/pdu_runtime.ex
defmodule HakoPdu.Runtime do
  @pdu_meta_data_size 24
  @pdu_meta_data_magicno 0x12345678
  @pdu_meta_data_version 1

  defmodule PduMetaDataState do
    def invalid, do: 0
    def uninitialized, do: 1
    def valid, do: 2
  end

  def classify_pdu_metadata(nil), do: PduMetaDataState.invalid()

  def classify_pdu_metadata(binary_data) when is_binary(binary_data) do
    cond do
      byte_size(binary_data) < @pdu_meta_data_size ->
        PduMetaDataState.invalid()

      binary_part(binary_data, 0, @pdu_meta_data_size) == :binary.copy(<<0>>, @pdu_meta_data_size) ->
        PduMetaDataState.uninitialized()

      true ->
        <<magic::little-unsigned-integer-size(32),
          version::little-unsigned-integer-size(32), _rest::binary>> = binary_data

        if magic == @pdu_meta_data_magicno && version == @pdu_meta_data_version do
          PduMetaDataState.valid()
        else
          PduMetaDataState.invalid()
        end
    end
  end

  def classify_pdu_metadata(_), do: PduMetaDataState.invalid()
end
